src/parsing/chunked_reader.rs

```diff
@@ -34,95 +32,64 @@ where
     pub fn new(reader: BufReader<R>) -> ChunkedReader<R> {
         ChunkedReader {
             inner: reader,
-            is_expecting_chunk: true,
-            read: 0,
-            length: 0,
-            line: Vec::new(),
+            buffer: Vec::new(),
+            consumed: 0,
+            reached_eof: false,
         }
     }
 
-    #[inline]
-    fn remaining(&self) -> u64 {
-        self.length - self.read
-    }
-
-    #[inline]
-    fn read_line(&mut self) -> io::Result<usize> {
-        buffers::read_line(&mut self.inner, &mut self.line)
-    }
-
-    fn read_chunk_size(&mut self) -> io::Result<u64> {
-        self.read_line()?;
-        if self.line.is_empty() {
+    fn read_chunk_size(&mut self) -> io::Result<usize> {
+        buffers::read_line(&mut self.inner, &mut self.buffer)?;
+        if self.buffer.is_empty() {
             return Err(io::ErrorKind::UnexpectedEof.into());
         }
-        parse_chunk_size(&self.line)
-    }
-
-    fn read_empty_line(&mut self) -> io::Result<()> {
-        let n = self.read_line()?;
-        if n == 0 || !self.line.is_empty() {
-            Err(InvalidResponseKind::Chunk.into())
-        } else {
-            Ok(())
-        }
+        parse_chunk_size(&self.buffer)
     }
 }
 
-impl<R> Read for ChunkedReader<R>
+impl<R> BufRead for ChunkedReader<R>
 where
     R: Read,
 {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if self.is_expecting_chunk {
-            debug!("waiting, parsing new chunk size");
-            // If we're waiting for a new chunk, we read a line and parse the number as hexadecimal.
-            self.read = 0;
-            self.length = self.read_chunk_size()?;
-            // If the chunk's length is 0, we've received the EOF chunk.
-            if self.length == 0 {
-                debug!("received EOF chunk");
-                // Read CRLF
-                self.read_empty_line()?;
-            }
-            self.is_expecting_chunk = false;
-        }
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        if self.buffer.len() == self.consumed && !self.reached_eof {
+            let chunk_size = self.read_chunk_size()?;
 
-        // If we have a length of 0 for a chunk, we've reached EOF.
-        if self.length == 0 {
-            return Ok(0);
-        }
+            self.buffer.resize(chunk_size + 2, 0);
```

Comment:
I'm a bit concerned by this, if a server sends large chunks it will cause the library to allocate a lot of memory. I think we can still implement BufRead without buffering the whole chunk at once.

Reply:
Yes, if we track how much of the chunk we still need to read the next time we have to refill the buffer and switch on whether we are mid-chunk or not.

But this does raise two question: What is a sensible amount of buffering? (For example the standard library internally goes for 8 kB buffers when e.g. `BufReader::new` is used.) And do want to try to always fill this buffer even if it means parsing multiple chunks which complicate things further?

Another approach might be to just limit the chunk size to e.g. 16 bit and hence 64 kB chunks.

Also note that we already have quite a few such issues in the code base, e.g. `read_chunk_size` will happily read a 3 GB buffer of digits until terminated by `\n` before it actually tries to parse this into number. So if we are not going to fix them all by limiting buffers everywhere, it would be unfortunate to loose the simplicity IMHO of buffering one chunk at a time.

Reply:
> And do want to try to always fill this buffer even if it means parsing multiple chunks which complicate things further?

I think we could just return a shorter buffer when the chunk is smaller the entire buffer.

> Also note that we already have quite a few such issues in the code base, e.g. read_chunk_size will happily read a 3 GB buffer of digits until terminated by \n before it actually tries to parse this into number. 

That's very true, I think it's something we might want to address eventually. It's probably more important for a server to limit the sizes of the data received because bad actors might try to DDOS it. I don't think that it's likely that a malicious server would try to break our library by announcing large chunks. I think that it's more likely that a server will unintentionally return large chunks and cause the library to use a lot of memory. I think we should put a hard cap on the buffer size at like 32 or 64 KiB.

Reply:
I added a commit that limits the chunk size to `u16` which implies a limit of 64 KB for the chunk buffer.

Reply:
So if a chunk larger than 64 KiB is sent we get an error? I'm not sure if I like that, I'd prefer just splitting the chunk in multiple calls to fill_buf.

Reply:
> I'd prefer just splitting the chunk in multiple calls to fill_buf.

I can certainly implement that but it is basically the same as implementing any arbitrary fixed amount of buffering which will complicate the implementation as it needs to keep state for handling partially read chunks. As said above, I am wary of the amount of ceremony if it does not buy us anything until e.g. `read_chunk_size` and friends are fixed.

Reply:
Pushed a variant that reads at most a single chunk and at most `MAX_BUFFER_LEN` bytes into the internal buffer. I set `MAX_BUFFER_LEN` to 64 kB but tested it using 3 B.